engine/schema/src/com/cloud/upgrade/DatabaseVersionHierarchy.java

```diff
@@ -0,0 +1,185 @@
+     * calculation assumes that the <code>fromVersion</code> required no schema migrations or data conversions
+     * and no  upgrade path was defined  for it.  Therefore, we find  the most recent  version  with  database
+     * migrations before the <code>fromVersion</code> and adopt that upgrade path list.
+     *
+     * @param fromVersion The version from which the upgrade will occur
+     * @param toVersion The version up to which the upgrade will occur
+     *
+     * @return The upgrade path from <code>fromVersion</code> to <code>toVersion</code>
+     */
+    public DbUpgrade[] getPath(final CloudStackVersion fromVersion, final CloudStackVersion toVersion) {
+        if (fromVersion == null) {
+            return new DbUpgrade[0];
+        }
+
+        // we cannot find the version specified, so get the
+        // most recent one immediately before this version
+        if (!contains(fromVersion)) {
+            return getPath(getRecentVersion(fromVersion));
+        }
+
+        final Predicate<? super VersionNode> predicate;
+
+        if (toVersion == null) {
+            // all the available versions greater than or equal to fromVersion
+            predicate = node -> node.version.compareTo(fromVersion) > -1;
+        } else {
+            // all the available versions greater than or equal to fromVersion AND less than toVersion
+            predicate = node -> node.version.compareTo(fromVersion) > -1 && node.version.compareTo(toVersion) < 0;
+        }
+
+        // get upgrade path from version forward (include version itself in the path)
+        return hierarchy
+                    .stream()
+                    .filter(predicate)
+                    .filter(distinct(node -> node.upgrader.getUpgradedVersion()))
+                    .map(node -> node.upgrader)
+                    .toArray(DbUpgrade[]::new);
+    }
+
+    /**
+     * Find the most recent <code>CloudStackVersion</code> immediately before <code>fromVersion</code>
+     *
+     * @param fromVersion The version to look up its immediate previous available version
+     *
+     * @return The <code>CloudStackVersion</code> or null
+     *
+     * @since 4.8.2.0 (refactored in 4.11.1.0)
+     */
+    private CloudStackVersion getRecentVersion(final CloudStackVersion fromVersion) {
+        if (fromVersion == null) {
+            return null;
+        }
+
+        // find the most recent version immediately before fromVersion
+        return toList()
+                .reverse()
+                .stream()
+                .filter(version -> fromVersion.compareTo(version) < 0)
+                .findFirst()
+                .orElse(null);
+    }
+
+    /**
+     * Generate immutable list of available <code>CloudstackVersion</code> in the hierarchy
+     *
+     * @return list of available versions
+     */
+    public ImmutableList<CloudStackVersion> toList() {
+        List<CloudStackVersion> versions = hierarchy
```

Comment:
You for sure like Lambda and Streams :)

Reply:
They are really useful, specially that it shortens LoC a lot! :)

Reply:
Yes, they are useful, but I still have not adopted them. I have some doubts regarding performance and readability with time. It is not a problem in this context that you are using them though.

Reply:
It has a getting used to side on it. But once you got used to it, it's super fun. Specially `FunctionalInterface`, `Consumer`, `Supplier` and friends.
Using lambda doesn't have any performance impact at all, but stream and specially filters might, if misused. (they are essentially a loop with time and space complexity around them anyway)